drv/ignition-server/src/main.rs

```diff
@@ -78,19 +85,126 @@ struct ServerImpl {
+    }
+
+    /// Apply the given function to each port for which a bit in the `ports`
+    /// vector is set. Returns a bit vector with bits set for ports for which
+    /// the operation was succesfull. Under normal circumstances this output
+    /// vector is expected to match the input vector.
+    fn map_ports<F>(&self, ports: u64, mut f: F) -> u64
+    where
+        F: FnMut(u8) -> Result<(), IgnitionError>,
+    {
+        let mut success = 0u64;
+
+        for port in 0..self.port_count.min(PORT_MAX) {
+            let mask = 1 << port;
+
+            if ports & mask != 0 {
+                match f(port) {
+                    Ok(()) => success |= mask,
+                    Err(e) => ringbuf_entry!(Trace::TargetError(port, e)),
+                }
+            }
+        }
+
+        success
+    }
+
+    /// Callback which gets called whenever a Target is first seen.
+    fn target_arrive(&self, port: u8) -> Result<(), IgnitionError> {
+        ringbuf_entry!(Trace::TargetArrive(port));
+
+        // Clear counters.
+        self.controller.counters(port)?;
+
+        // Reset the link events for each transceiver if the register is set to
+        // its default value.
+        for txr in &TransceiverSelect::ALL {
+            let events = LinkEvents::from(
+                self.controller
+                    .link_events(port, *txr)
+                    .map_err(IgnitionError::from)?,
+            );
+
+            if events == LinkEvents::ALL {
+                self.controller
+                    .clear_link_events(port, *txr)
+                    .map_err(IgnitionError::from)?;
+            }
         }
 
         Ok(())
     }
 
-    fn port_state(&self, port: u8) -> Result<PortState, IgnitionError> {
-        self.controller.state(port).map_err(IgnitionError::from)
+    /// Callback which gets called whenever a Target goes away.
+    fn target_depart(&self, port: u8) -> Result<(), IgnitionError> {
+        ringbuf_entry!(Trace::TargetDepart(port));
+        Ok(())
+    }
+
+    fn target_request(
+        &self,
+        port: u8,
+        request: Request,
+    ) -> Result<(), IgnitionError> {
+        if self.target(port)?.request_in_progress() {
+            return Err(IgnitionError::RequestInProgress);
+        }
+
+        self.controller
+            .set_request(port, request)
+            .map_err(IgnitionError::from)?;
+
+        // Determine if the request was accepted by matching the (updated)
+        // Target power state with the request.
+        match (request, self.target(port)?.power_state) {
+            (Request::SystemPowerOff, SystemPowerState::PoweringOff)
+            | (Request::SystemPowerOff, SystemPowerState::Off)
+            | (Request::SystemPowerOn, SystemPowerState::PoweringOn)
+            | (Request::SystemPowerOn, SystemPowerState::On)
+            | (Request::SystemPowerReset, SystemPowerState::PoweringOff)
+            | (Request::SystemPowerReset, SystemPowerState::PoweringOn) => {
```

Comment:
I'm on the fence whether this would be an improvement, but patterns can be nested if it would be clearer to only have one arm per `Request` variant:

```rust
            (
                Request::SystemPowerOff,
                SystemPowerState::PoweringOff | SystemPowerState::Off,
            )
            | (
                Request::SystemPowerOn,
                SystemPowerState::PoweringOn | SystemPowerState::On,
            )
            | (
                Request::SystemPowerReset,
                SystemPowerState::PoweringOff | SystemPowerState::PoweringOn,
            ) => Ok(()),
```

Reply:
Sure, I had this in an earlier iteration. It ends up visually grouping these, which is may aid understanding. I don't feel strongly one way or another.

Reply:
I don't either. Leaving it the way it is is fine with me.